kroxylicious-additional-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/EnvelopeEncryptionFilter.java

```diff
@@ -0,0 +1,228 @@
+            MemoryRecordsBuilder builder = recordsBuilder(buffer, records);
+            for (var kafkaRecord : records.records()) {
+                ByteBuffer output = null;
+                try {
+                    output = bufferPool.acquire(dekContext.encodedSize(kafkaRecord.valueSize()));
+                    dekContext.encode(kafkaRecord.value(), output);
+                    builder.append(kafkaRecord.timestamp(), kafkaRecord.key(), output, kafkaRecord.headers());
+                }
+                finally {
+                    if (output != null) {
+                        bufferPool.release(output);
+                    }
+                }
+            }
+            newPartitionData.add(new PartitionProduceData()
+                    .setIndex(partitionId)
+                    .setRecords(builder.build()));
+        }
+        return newPartitionData;
+    }
+
+    @Override
+    public CompletionStage<ResponseFilterResult> onFetchResponse(short apiVersion, ResponseHeaderData header, FetchResponseData response, FilterContext context) {
+        return maybeDecodeFetch(response.responses())
+                .thenCompose(responses -> context.forwardResponse(header, new FetchResponseData().setResponses(responses)));
+    }
+
+    private CompletableFuture<List<FetchableTopicResponse>> maybeDecodeFetch(List<FetchableTopicResponse> topics) {
+        List<CompletableFuture<FetchableTopicResponse>> result = new ArrayList<>();
+        for (FetchableTopicResponse topicData : topics) {
+            result.add(maybeDecodePartitions(topicData.partitions()).thenApply(kk -> {
+                topicData.setPartitions(kk);
+                return topicData;
+            }));
+        }
+        return join(result);
+    }
+
+    private CompletableFuture<List<PartitionData>> maybeDecodePartitions(List<PartitionData> partitions) {
+        List<CompletableFuture<PartitionData>> result = new ArrayList<>();
+        for (PartitionData partitionData : partitions) {
+            if (!(partitionData.records() instanceof MemoryRecords)) {
+                throw new IllegalStateException();
+            }
+            result.add(maybeDecodeRecords(partitionData, (MemoryRecords) partitionData.records()));
+        }
+        return join(result);
+    }
+
+    private CompletableFuture<PartitionData> maybeDecodeRecords(PartitionData fpr, MemoryRecords memoryRecords) {
+        final CompletableFuture<PartitionData> result;
+        if (!anyRecordSmellsOfEncryption(memoryRecords)) {
+            result = CompletableFuture.completedFuture(fpr);
+        }
+        else {
+            ByteBuffer buffer = null; // TODO
+            MemoryRecordsBuilder builder = recordsBuilder(buffer, memoryRecords);
+            List<CompletableFuture<Integer>> futures = new ArrayList<>();
+            for (var kafkaRecord : memoryRecords.records()) {
+                if (!smellsOfEncryption(kafkaRecord.value())) {
```

Comment:
Think we need to special case records with null values.

Reply:
Yes, we should encrypt null values just like any other. Added a TODO